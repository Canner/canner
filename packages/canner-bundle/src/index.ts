import webpack from 'webpack';
import WebpackDevServer from 'webpack-dev-server';
import {createConfig, CreateConfigArgsType, createWebConfig, createSchemaConfig} from './utils/createWebpackConfig';
import transformSchemaToJson from './utils/transformSchemaToJSON';
import {SCHEMA_OUTPUT_PATH, SCHEMA_JSON_OUTPUT_PATH} from './config';
export function build(options?: CreateConfigArgsType) {
  const config = createConfig(options || {});
  return new Promise((resolve, reject) => {
    webpack(config)
      .run((err: any, stats) => {
        if (err) {
          console.error(err.stack || err);
          if (err.details) {
            return reject(err.details);
          }
        }
      
        const info = stats.toJson();
      
        if (stats.hasErrors()) {
          return reject(info.errors);
        }
        if (!options.webOnly) {
          const schemaPath = `${(options || {}).schemaOutputPath || SCHEMA_OUTPUT_PATH}/schema.node.js`;
          transformSchemaToJson(schemaPath, (options || {}).schemaJsonOutputPath || SCHEMA_JSON_OUTPUT_PATH);
        }
        resolve(stats);
      });
  });
}

export function watchSchema(options: CreateConfigArgsType, callback: any) {
  if (options && !options.watch) {
    options.watch = true;
  }
  const config = createSchemaConfig(options || {});
  const watchOptions = typeof config.watch === 'object'?
    config.watch :
    {
      aggregateTimeout: 1000,
    };
  return webpack(config)
    .watch(watchOptions, (err, stats) => {
      try {
        const schemaPath = `${(options || {}).schemaOutputPath || SCHEMA_OUTPUT_PATH}/schema.node.js`;
        transformSchemaToJson(schemaPath, (options || {}).schemaJsonOutputPath || SCHEMA_JSON_OUTPUT_PATH);
      } catch(e) {
        err = e;
      }
      callback(err, stats);
    });
}

export function createWebpackDevServer(options: CreateConfigArgsType) {
  if (options) {
    options.watch = true;
  }
  const config = createWebConfig(options || {}) as any;
  config.entry.index.unshift(`webpack-dev-server/client?http://localhost:${config.devServer.port}/`, `webpack/hot/dev-server`);
  const compiler = webpack(config);
  const devServerOptions = {...config.devServer};
  WebpackDevServer.addDevServerEntrypoints(config, devServerOptions);
  const server = new WebpackDevServer(compiler, devServerOptions);
  return {
    compiler,
    server,
    config
  };
}
