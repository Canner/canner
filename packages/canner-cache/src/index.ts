import {CacheI, Action, Listeners, Reducer} from './types';

export default class Cache implements CacheI {
  reducer: (data: any, action: Action) => any;
  data: Record<string, any>;
  listeners: Listeners;
  constructor({reducer}: {reducer: Reducer}) {
    this.reducer = (data, actions) => {
      // action array
      if (Array.isArray(actions)) {
        return actions.reduce(reducer, data);
      }
      // single action
      return reducer(data, actions);
    };
    this.data = {};
    this.listeners = {}
  }

  isCached = key => {
    // check the key is cached or not
    return Boolean(this.data[key]);
  };

  setData = (key, data) => {
    this.data[key] = data;
    this.publish(key);
  }

  getData = key => {
    if (this.isCached(key)) {
      return this.data[key];
    }
    throw new Error(`There is no cached data with key '${key}'`);
  };

  removeData = (key: string) => {
    delete this.data[key];
    delete this.listeners[key];
  }

  mutate = (key, actions) => {
    const originData = this.getData(key);
    const newData = this.reducer(originData, actions);
    this.setData(key, newData);
  }

  subscribe = (key, callback) => {
    const listener = {
      id: randomId(),
      callback
    };
    if (key in this.listeners) {
      this.listeners[key].push(listener) - 1;
    } else {
      this.listeners[key] = [listener];
    }
    return listener.id;
  }

  unsubscribe = (key, id) => {
    if (this.listeners[key]) {
      this.listeners[key] = this.listeners[key].filter(listener => listener.id !== id);
    }
  }

  publish = (key) => {
    const data = this.data[key];
    if (key in this.listeners) {
      this.listeners[key].forEach(listener => {
        listener.callback(data);
      });
    }
  }
}

function randomId(): string {
  return Math.random().toString(36).replace(/[^a-z]+/g, '').substr(2, 10);
}