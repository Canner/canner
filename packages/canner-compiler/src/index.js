import Traverser from './traverser';
import Parser from './parser';

export default {
  Traverser,
  Parser,
};
