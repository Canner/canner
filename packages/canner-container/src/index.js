export default, {transformSchemaToMenuConfig } from './components';
export { default as Sidebar } from './components/Sidebar';
export { default as Navbar } from './components/Navbar';
