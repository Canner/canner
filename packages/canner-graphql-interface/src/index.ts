import { createLink } from './link';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache, NormalizedCacheObject } from 'apollo-cache-inmemory';
import { Connector } from './connector/types';
import { CustomizeResolver } from './resolver';
import { GraphqlClient } from './graphqlClient/types';
import FirebaseRtdbClientConnector from './connector/firebaseConnector';
import FirestoreClientConnector from './connector/firestoreConnector';
import FirestoreAdminConnector from './connector/firestoreAdminConnector';
import MemoryConnector from './connector/memoryConnector';
import LocalStorageConnector from './connector/localStorageConnector';
import CannerPlatformConnector from './connector/cannerPlatformConnector';
import PrismaClient from './graphqlClient/prismaClient';
import GraphqlClientImpl from './graphqlClient/graphqlClient';

class FirebaseRtdbAdminConnector extends CannerPlatformConnector {
  constructor({namespace, projectId}: {namespace?: string; databaseURL: string; projectId: string}) {
    super();
    this.service = 'firebase';
    this.resourceId = projectId;
    this.connectorName = 'firebase.rtdb.admin';
    this.connectorParams = {
      namespace
    };
  }
}

const createClient = ({
  schema,
  connector,
  connectors,
  resolvers,
  graphqlClient,
}: {
  schema: any,
  connectors?: {[key: string]: Connector};
  resolvers?: {[key: string]: CustomizeResolver};
  graphqlClient?: GraphqlClient;
  connector?: Connector;
}) => {
  const apolloCache = new InMemoryCache();
  return new ApolloClient({
    cache: apolloCache,
    link: createLink({schema, connectors, graphqlClient, connector, resolvers})
  });
};

export {
  createClient,
  MemoryConnector,
  LocalStorageConnector,
  FirebaseRtdbAdminConnector,
  FirebaseRtdbClientConnector,
  FirestoreAdminConnector,
  FirestoreClientConnector,
  PrismaClient,
  GraphqlClientImpl as GraphqlClient
};
