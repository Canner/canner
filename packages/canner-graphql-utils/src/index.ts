import { filter } from './filter';
import paginator from './paginator';
import { Connector } from './types';

export {
  filter,
  paginator,
  Connector
};
