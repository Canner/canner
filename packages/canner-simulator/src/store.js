// @flow

// $FlowFixMe: unresolved
import {fromJS} from 'immutable';
import RefId from 'canner-ref-id';

export type UpdateType = 'update' | 'create' | 'delete';
export type SwapType = 'swap';

export type onChangeFunc =  ((path: RefId, type: UpdateType, delta: *) => void) &
  ((path: {firstRefId: RefId, secondRefId: RefId}, type: SwapType) => void);

class Store {
  // eslint-disable-next-line
  subcriber = (value: *) => {};
  value = fromJS({});

  getValue = () => {
    return this.value;
  }

  listen(subcriber: (value: *) => void) {
    this.subcriber = subcriber;
  }

  push() {
    this.subcriber(this.value);
  }

  setValue = (value: *) => {
    this.value = fromJS(value);
    this.push();
  }

  getValueByPath = (path: RefId) => {
    const pathArr = path.getPathArr();
    return this.value.getIn(pathArr);
  }

  setValueByPath = (path: RefId, value: *) => {
    const pathArr = path.getPathArr();
    this.value = this.value.setIn(pathArr, fromJS(value));
    this.push();
  }

  // $FlowFixMe: flow can't work pretty on function overloading , so using any to temp resolve
  onChange: onChangeFunc = (path: any, type, delta: any) => {
    switch(type) {
      case 'update': {
        this.setValueByPath(path, delta);
        break;
      }
      case 'create': {
        let collection = this.getValueByPath(path);
        if (isEntry(path)) {
          delta = delta.set('_id', getRandomString());
        }
        collection = collection.push(delta);
        this.setValueByPath(path, collection);
        break;
      }
      case 'delete': {
        let collection = this.getValueByPath(path.remove());
        const pathArr = path.getPathArr();
        const index = pathArr[pathArr.length - 1];
        collection = collection.remove(index);
        this.setValueByPath(path.remove(), collection);
        break;
      }
      case 'swap': {
        const {firstRefId, secondRefId} = path;
        const collection = this.getValueByPath(firstRefId.remove());
        const firstRefIdArr = firstRefId.getPathArr();
        const secondRefIdArr = secondRefId.getPathArr();
        const firstIndex = firstRefIdArr[firstRefIdArr.length - 1];
        const secondIndex = secondRefIdArr[secondRefIdArr.length - 1];
        let newCollection = collection.set(firstIndex, collection.get(secondIndex));
        newCollection = newCollection.set(secondIndex, collection.get(firstIndex));
        this.setValueByPath(firstRefId.remove(), newCollection);
        break;
      }
      default:
        break;
    }
  }
}

function isEntry(path: RefId) {
  return path.getPathArr.length === 1;
}

function getRandomString() {
  return Math.random().toString(36).substr(2, 6);
}

export default new Store();