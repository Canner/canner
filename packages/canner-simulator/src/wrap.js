// @flow

import * as React from 'react';
// eslint-disable-next-line
import {Button, message} from 'antd';
import { Context, createEmptyData } from 'canner-helpers';
import RefId from 'canner-ref-id';
import store from './store';
import type { onChangeFunc, UpdateType } from './store';

type Props = {
  value: *,
  refId?: RefId,
  keyName: string,
  items?: Object,
  onChange?: onChangeFunc,
  className?: string,
  children: typeof React.Children,
  liteCMS?: React.ComponentType<*>,
  style?: Object,
  title?: string
};

type State = {
  value: *,
  refId: RefId
};

const wrap = (Com: React.ComponentType<*>) => class wrappedComponent extends React.PureComponent<Props, State> {
  static defaultProps = {
    refId: null,
    items: {},
    onChange: () => Promise.resolve(),
    className: undefined,
    liteCMS: null,
    style: {},
    title: ''
  }

  constructor(props: Props) {
    super(props);
    const refId = props.refId
      ? props.refId.child(props.keyName)
      : new RefId(props.keyName);
    this.state = {
      value: getValue(store.getValue(), refId),
      refId,
    };
  }

  componentWillReceiveProps() {
    this.updateValue();
  }

    onChange = (refId: RefId, type: UpdateType, delta: *) => {
      let cpDelta = delta;
      const { items = {}, onChange } = this.props;
      if (onChange) {
        onChange(refId, type, delta);
      }

      if (type === 'create') {
        cpDelta = createEmptyData(items);
      }

      store.onChange(refId, type, cpDelta);
      this.updateValue();
    }

    updateValue = () => {
      this.setState(prevState => ({
        value: getValue(store.getValue(), prevState.refId),
      }));
    }

    render() {
      const { value, refId } = this.state;
      const {
        children, title, liteCMS, className, style = {}, keyName, ...rest
      } = this.props;
      const contextValue = {
        renderChildren: props => (children ? React.Children.map(children, child => React.cloneElement(child, props)) : 'this is content'),
        renderComponent: props => (liteCMS ? <liteCMS {...props} /> : 'this is content'),
        renderConfirmButton: genFakeButton({ text: 'Confirm' }),
        renderCancelButton: genFakeButton({ text: 'Cancel' }),
        refId,
      };
      return (
        <Context.Provider value={contextValue}>
          <div
            className={className}
            style={{
              marginTop: 16,
              ...style,
            }}
          >
            <h3>
              {title || keyName.toUpperCase()}
            </h3>
            <Com
              {...rest}
              keyName={keyName}
              onChange={this.onChange}
              value={value}
              refId={refId}
              goTo={goTo}
            />
          </div>
        </Context.Provider>
      );
    }
};
export default wrap;

function getValue(value, refId) {
  const pathArr = refId.getPathArr() || [];
  return value.getIn(pathArr);
}

function goTo(path) {
  message.success(`go to 'baseUrl/${path}`);
}

function genFakeButton({
  // eslint-disable-next-line
  text,
  // eslint-disable-next-line
  style
}: Object) {
  return function Button({
    disabled = false,
    // eslint-disable-next-line
    style = style,
    // eslint-disable-next-line
    refId,
    // eslint-disable-next-line
    onClick,
    // eslint-disable-next-line
    callback = () => {},
    text = text,
    component = Button,
  } = {}) {
    return React.createElement(component, {
      disabled,
      style,
      onClick: () => {},
    }, text);
  };
}
