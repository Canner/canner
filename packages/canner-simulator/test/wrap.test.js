import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import RefId from 'canner-ref-id';
import { Item, LiteCMS } from 'canner-helpers';
import { store, wrap } from '../src';

Enzyme.configure({ adapter: new Adapter() });

const mockArrayComponent = ({ refId }) => (
  <div>
    component1
    <Item refId={refId.child(0)} />
  </div>
);

const mockObjectComponent = () => (
  <Item />
);

const mockStringComponent = ({ value, onChange, refId }) => (
  <input value={value} onChange={e => onChange(refId, 'update', e.target.value)} />
);

const mockLiteCMSComponent = () => (
  <LiteCMS />
);

const WrappedArrayComponent = wrap(mockArrayComponent);
const WrappedObjectComponent = wrap(mockObjectComponent);
const WrappedStringComponent = wrap(mockStringComponent);
const WrappedLiteCMSComponent = wrap(mockLiteCMSComponent);


describe('wrap simple', () => {
  it('should has props: refId, value and onChange', () => {
    store.setValue({
      info: {
        test: '123',
      },
    });

    const wrapper = mount(<WrappedStringComponent keyName="info/test" />);
    expect(wrapper.find('input').prop('value')).toBe('123');
    wrapper.find('input').simulate('change', { target: { value: '321' } });
    expect(wrapper.find('input').prop('value')).toBe('321');
  });

  it('should support onChange to listen', () => {
    store.setValue({
      info: {
        test: '123',
      },
    });
    const mockOnChange = jest.fn();
    const wrapper = mount(<WrappedStringComponent keyName="info/test" onChange={mockOnChange} />);
    wrapper.find('input').simulate('change', { target: { value: '321' } });
    expect(mockOnChange).toBeCalledWith(expect.any(RefId), 'update', '321');
  });

  it('should support item', () => {
    store.setValue({
      info: {
        test: '123',
      },
    });

    const wrapper = mount(<WrappedObjectComponent keyName="info/test" />);
    expect(wrapper.find(mockObjectComponent).children().length).toBe(1);
  });

  it('should support customize item', () => {
    store.setValue({
      info: {
        test: '123',
      },
    });

    const MyItem = jest.fn().mockImplementation(() => 'my item');
    const wrapper = mount(
      <WrappedObjectComponent keyName="info/test">
        <MyItem />
      </WrappedObjectComponent>
    );
    expect(MyItem.mock.calls[0][0]).toEqual({
      refId: expect.any(RefId),
    });
    expect(wrapper.find(mockObjectComponent).text()).toBe(MyItem());
  });

  it('should support liteCMS', () => {
    store.setValue({
      info: {
        test: '123',
      },
    });
    const wrapper = mount(<WrappedLiteCMSComponent keyName="info/test" />);
    expect(wrapper.find(mockLiteCMSComponent).children().length).toBe(1);
  });
});

describe('wrap comlex', () => {
  it('should support customize item get value', () => {
    store.setValue({
      info: {
        test: '123',
      },
    });
    const wrapper = mount(
      <WrappedObjectComponent
        keyName="info"
      >
        <WrappedStringComponent
          keyName="test"
        />
      </WrappedObjectComponent>
    );
    expect(wrapper.find('input').prop('value')).toBe('123');
    wrapper.find('input').simulate('change', { target: { value: '321' } });
    expect(wrapper.find('input').prop('value')).toBe('321');
  });

  it('should support customize array item get value', () => {
    store.setValue({
      posts: [{
        test: '123',
      }],
    });
    const wrapper = mount(
      <WrappedArrayComponent
        keyName="posts"
      >
        <WrappedStringComponent
          keyName="test"
        />
      </WrappedArrayComponent>
    );
    expect(wrapper.find('input').prop('value')).toBe('123');
    wrapper.find('input').simulate('change', { target: { value: '321' } });
    expect(wrapper.find('input').prop('value')).toBe('321');
  });

  it('should support customize complex item', () => {
    store.setValue({
      users: [{
        basicInfo: {
          name: '123',
          nickname: '321',
        },
        addresses: [{
          type: 'home',
          value: 'xxxxxx',
        }],
        username: 'haha',
      }],
    });
    const wrapper = mount(
      <WrappedArrayComponent
        keyName="users"
      >
        <WrappedObjectComponent keyName="basicInfo">
          <WrappedStringComponent keyName="name" />
          <WrappedStringComponent keyName="nickname" />
        </WrappedObjectComponent>
        <WrappedArrayComponent keyName="addresses">
          <WrappedStringComponent keyName="type" className="type" />
          <WrappedStringComponent keyName="value" className="value" />
        </WrappedArrayComponent>
        <WrappedStringComponent keyName="username" />
      </WrappedArrayComponent>
    );
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.find('.type input')
      .simulate('change', { target: { value: 'company' } });
    wrapper.find('.value input')
      .simulate('change', { target: { value: 'yyyyy' } });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
