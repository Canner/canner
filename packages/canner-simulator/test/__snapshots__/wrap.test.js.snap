// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrap comlex should support customize complex item 1`] = `"<div style=\\"margin-top: 16px;\\"><h3>USERS</h3><div>component1<div style=\\"margin-top: 16px;\\"><h3>BASICINFO</h3><div style=\\"margin-top: 16px;\\"><h3>NAME</h3><input value=\\"123\\"></div><div style=\\"margin-top: 16px;\\"><h3>NICKNAME</h3><input value=\\"321\\"></div></div><div style=\\"margin-top: 16px;\\"><h3>ADDRESSES</h3><div>component1<div class=\\"type\\" style=\\"margin-top: 16px;\\"><h3>TYPE</h3><input value=\\"home\\"></div><div class=\\"value\\" style=\\"margin-top: 16px;\\"><h3>VALUE</h3><input value=\\"xxxxxx\\"></div></div></div><div style=\\"margin-top: 16px;\\"><h3>USERNAME</h3><input value=\\"haha\\"></div></div></div>"`;

exports[`wrap comlex should support customize complex item 2`] = `"<div style=\\"margin-top: 16px;\\"><h3>USERS</h3><div>component1<div style=\\"margin-top: 16px;\\"><h3>BASICINFO</h3><div style=\\"margin-top: 16px;\\"><h3>NAME</h3><input value=\\"123\\"></div><div style=\\"margin-top: 16px;\\"><h3>NICKNAME</h3><input value=\\"321\\"></div></div><div style=\\"margin-top: 16px;\\"><h3>ADDRESSES</h3><div>component1<div class=\\"type\\" style=\\"margin-top: 16px;\\"><h3>TYPE</h3><input value=\\"company\\"></div><div class=\\"value\\" style=\\"margin-top: 16px;\\"><h3>VALUE</h3><input value=\\"yyyyy\\"></div></div></div><div style=\\"margin-top: 16px;\\"><h3>USERNAME</h3><input value=\\"haha\\"></div></div></div>"`;
